syncer/syncer.go

```diff
@@ -1305,22 +1305,36 @@ func (s *Syncer) syncDML(
 		jobs = jobs[0:0]
 	}
 
-	// successF is used to calculate lag metric and tps.
+	// successF is used to calculate lag metric and q/tps.
 	successF := func() {
 		if len(jobs) > 0 {
 			// NOTE: we can use the first job of job queue to calculate lag because when this job committed,
 			// every event before this job's event in this queue has already commit.
 			// and we can use this job to maintain the oldest binlog event ts among all workers.
 			j := jobs[0]
 			s.updateReplicationLag(j, workerLagKey)
+
+			// metric only increases by 1 because dm batches sql jobs in a single transaction.
+			metrics.FinishedTransactionTotal.WithLabelValues(s.cfg.Name, s.cfg.WorkerName, s.cfg.SourceID).Inc()
+			switch j.tp {
+			case ddl:
+				metrics.BinlogEventCost.WithLabelValues(metrics.BinlogEventCostStageDDLExec, s.cfg.Name, s.cfg.WorkerName, s.cfg.SourceID).Observe(time.Since(j.jobAddTime).Seconds())
+			case insert, update, del:
+				metrics.BinlogEventCost.WithLabelValues(metrics.BinlogEventCostStageDMLExec, s.cfg.Name, s.cfg.WorkerName, s.cfg.SourceID).Observe(time.Since(j.jobAddTime).Seconds())
+			}
 		} else {
 			s.updateReplicationLag(nil, workerLagKey)
 		}
-		// calculate tps
-		for tpName, v := range tpCnt {
-			s.addCount(true, queueBucket, tpName, v)
-			tpCnt[tpName] = 0
+		// calculate qps
+		for dbSchema, tableM := range tpCnt {
+			for dbTable, tpM := range tableM {
+				for tpName, cnt := range tpM {
+					tpCnt[dbSchema][dbTable][tpName] = 0
```

Comment:
maybe we can delete every value in `tpCnt`, in your implementation the map will only grow 🤔 (Although I think upstream should not have such many tables)

Or at lease dont call `addCount` for `cnt == 0`

Reply:
maybe  i can clear `tpCnt` map in `clearF` ？ https://github.com/pingcap/dm/pull/1778/commits/2a6bfabb3dd120b087807f561ef1ae6a6d9f0798

Reply:
because tpCnt is reset in `clearF`, it's no need to assign every value to 0 and check `if cnt > 0` now